environment/lib/borsh/index.js

```diff
@@ -193,27 +194,79 @@ function deserialize (schema, fieldType, buffer) {
+    let errorMsg;
+    for (let i = 0; i < RETRY_SEND_TX; i++) {
+        let sendTxnAsync;
+        let txHash;
+        let resendLast = false;
 
-    const result = await account.connection.provider.sendTransaction(signedTx);
+        try {
+            if (resendLast) {
+                console.log('resend txn');
+                await sendTxnAsync();
+                resendLast = false;
+            } else {
+                const status = await account.connection.provider.status();
+                let signedTx;
+                [txHash, signedTx] = await nearlib.transactions.signTransaction(
+                    receiverId, ++accessKey.nonce, actions, nearlib.utils.serialize.base_decode(status.sync_info.latest_block_hash), account.connection.signer, account.accountId, account.connection.networkId,
+                );
+                const bytes = signedTx.encode();
+                sendTxnAsync = async () => {
+                    await account.connection.provider.sendJsonRpc('broadcast_tx_async', [Buffer.from(bytes).toString('base64')]);
+                    console.log('TxHash', nearlib.utils.serialize.base_encode(txHash));
+                }
+                await sendTxnAsync();
+            }
+        } catch (e) {
+            // sleep to avoid socket hangout on retry too soon
+            await sleep(500);
+            continue;
+        }
 
-    const flatLogs = [result.transaction_outcome, ...result.receipts_outcome].reduce((acc, it) => acc.concat(it.outcome.logs), []);
-    if (flatLogs) {
-        console.log(flatLogs);
-    }
+        let result;
+        for (let j = 0; j < RETRY_TX_STATUS; j++) {
+            try {
+                result = await account.connection.provider.txStatus(txHash, account.accountId);
+                if (result.status.SuccessValue !== undefined || result.status.Failure !== undefined) {
+                    break;
+                }
+            } catch (e) {
+                await sleep((j+1)*500);
+            }
+        }
 
-    if (result.status.Failure) {
-        throw new Error(JSON.stringify(result.status.Failure));
+        if (result) {
+            const flatLogs = [result.transaction_outcome, ...result.receipts_outcome].reduce((acc, it) => acc.concat(it.outcome.logs), []);
+            if (flatLogs) {
+                console.log(flatLogs);
+            }
+    
+            if (result.status.SuccessValue !== undefined) {
+                return result;
+            }
+    
+            errorMsg = JSON.stringify(result.status.Failure);
+            if (errorMsg.includes('Transaction nonce')) {
+                // nonce incorrect, re-fetch nonce and retry
+                continue;
+            } else {
+                // Indeed txn error, retry doesn't help
+                break;
+            }
+        } else {
+            // Still no result after a long time, resubmit txn
+            resendLast = true;
+            continue;
+        }
```

Comment:
can only come up with quite lengthy way to ensure txn successfully handled by nodes, hopes we can have such a reliable way in near-api-js

Reply:
should be handled properly in 0.27.0

Reply:
Hmm just checked. Timeout is caught in 0.27.0, but these three cases are not:
- socket hangup
- retrying on correct nonce, in case submit with same key concurrently, or large nonce reach/process first, smaller nonce txn retry with correct nonce
- broadcast_txn wait seconds sometimes, cause socket hangup will not get a txn hash and also do not not txn is submit or not. Sometimes txn actually submitted. Then retry submit txn cause double txn send: https://buildkite.com/nearprotocol/rainbow-bridge/builds/283#9ca9cc8d-ca24-4f12-9954-bbfdba6a4dd1
I think expose broadcast_txn_async in near-api-js fix this (here i'm using jsonrpc to broadcast_txn_async), because this get response txnhash immediately, and we know it submitted, if we can get the response, then we just retry txn status (Just found this has exported) until get success or error.

Also Bowen mentioned even broadcast_txn_async and get a txn hash, due to our txn pool design it's every txn takes random long time to be picked up, it's possible some txn takes very long to going through, is this still the case? @bowenwang1996. Based on devnet load test, some takes even more than 100s, so in that case i resend txn with same nonce, and hopefully it has a larger chance to pickup, which has same nonce and won't be double processed.
cc @vgrichina

Reply:
> it's possible some txn takes very long to going through

that might still happen, but very unlikely.

Reply:
Thanks. In near-relayer as long running process rare likelihood could add up (or when there's burst in node's load), i'll still handle this case

Reply:
We have pinned nearcore version in our prepare.sh script. Created an issue: https://github.com/near/rainbow-bridge/issues/161

Reply:
@nearmax Can you also approve this. It doesn't fail on contract call nonce, near&web3 timeout and rpc errors, check reinit a near contract in the past 6 runs. It fails once on deploy contract call nonce, but probably would solve in #161.